python_modules/dagster/dagster/_core/execution/plan/execute_step.py

```diff
@@ -859,18 +895,41 @@ def _log_asset_materialization_events_for_asset(
             f"Unexpected asset execution type {execution_type}",
         )
 
-        yield from (
-            (
-                DagsterEvent.asset_materialization(step_context, materialization)
-                for materialization in _get_output_asset_materializations(
-                    asset_key,
-                    partitions,
-                    output,
-                    output_def,
-                    manager_metadata,
-                    step_context,
+        # This is a temporary workaround to prevent duplicate observation events from external
+        # observable assets that were auto-converted from source assets. These assets yield
+        # observation events through the context in their body, and will continue to do so until we
+        # can convert them to using ObserveResult, which requires a solution to partition-scoped
+        # metadata and data version on output. We identify these auto-converted assets by looking
+        # for OBSERVATION-type asset that have this special metadata key (added in
+        # `wrap_source_asset_observe_fn_in_op_compute_fn`), which should only occur for these
+        # auto-converted source assets. This can be removed when source asset observation functions
+        # are converted to use ObserveResult.
+        if (
+            execution_type == AssetExecutionType.OBSERVATION
+            and SYSTEM_METADATA_KEY_SOURCE_ASSET_OBSERVATION in output.metadata
+        ):
+            pass
+        elif execution_type != AssetExecutionType.UNEXECUTABLE:
+            yield from (
+                (
+                    _dagster_event_for_asset_event(step_context, event)
+                    for event in _get_output_asset_events(
+                        asset_key,
+                        partitions,
+                        output,
+                        output_def,
+                        manager_metadata,
+                        step_context,
+                        execution_type,
+                    )
                 )
             )
-            if execution_type == AssetExecutionType.MATERIALIZATION
-            else ()
-        )
+
```

Comment:
Would strongly prefer to have this structured so that there is no `elif` to demonstrate to the code reader that under no circumstances will these blocks be silently skipped.. The check against `UNEXECUTABLE` in `elif` is unnecessary given the invariant at top of function.

If you still want to check against unexecutable down here, do so as an invariant.
```
else:
   check.invariant(execution_type != AssetExecutionType.UNEXECUTABLE)
   yield from (...)
```